r/R/pkg-arrow.R

```diff
@@ -0,0 +1,168 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# exported in zzz.R
+infer_type.nanoarrow_array <- function(x, ...) {
+  arrow::as_data_type(infer_nanoarrow_schema(x, ...))
+}
+
+as_data_type.nanoarrow_schema <- function(x, ...) {
+  exportable_schema <- nanoarrow_allocate_schema()
+  nanoarrow_pointer_export(x, exportable_schema)
+  asNamespace("arrow")$DataType$import_from_c(exportable_schema)
```

Comment:
Not exported from arrow? We should fix that. 

```suggestion
  getFromNamespace("DataType", "arrow")$import_from_c(exportable_schema)
```